### PR TITLE
Cctx merge

### PR DESCRIPTION
**Pull request by Cyan4973**
`CCtx` and `CStream` are now same object.

Updated functions : 
`ZSTD_freeCCtx()` (and `ZSTD_freeCStream()`)
`ZSTD_sizeof_CCtx()` (and `ZSTD_sizeof_CStream()`)
`ZSTD_estimateCCtx()` (with a limitation, see fc5145955a0c53a9157579ce9dc2abcadeff5e32)

### CODE REVIEW

File: lib/compress/zstd_compress.c [hunk 1]
```diff
@@ -3124,26 +3145,23 @@ ZSTD_CStream* ZSTD_createCStream_advanced(ZSTD_customMem customMem)
     if (zcs==NULL) return NULL;
     memset(zcs, 0, sizeof(ZSTD_CStream));
     memcpy(&zcs->customMem, &customMem, sizeof(ZSTD_customMem));
-    zcs->cctx = ZSTD_createCCtx_advanced(customMem);
```
Comment (terrelln) [inline lib/compress/zstd_compress.c]: Can it be implemented like this?
```c
ZSTD_CStream* ZSTD_createCStream_advanced(ZSTD_customMem customMem)
{
    return ZSTD_createCCtx_advanced(customMem);
}
```

File: lib/compress/zstd_compress.c [hunk 2]
```diff
@@ -3124,26 +3145,23 @@ ZSTD_CStream* ZSTD_createCStream_advanced(ZSTD_customMem customMem)
...
+
+size_t ZSTD_estimateCStreamSize(ZSTD_compressionParameters cParams)
+{
+    size_t const CCtxSize = ZSTD_estimateCCtxSize(cParams);
+    size_t const blockSize = MIN(ZSTD_BLOCKSIZE_ABSOLUTEMAX, (size_t)1 << cParams.windowLog);
+    size_t const inBuffSize = ((size_t)1 << cParams.windowLog) + blockSize;
+    size_t const outBuffSize = ZSTD_compressBound(blockSize) + 1;
+    size_t const streamingSize = inBuffSize + outBuffSize;
+
+    return sizeof(ZSTD_CCtx) + CCtxSize + streamingSize;
```
Comment (terrelln) [inline lib/compress/zstd_compress.c]: `ZSTD_estimateCCtxSize()` already includes `sizeof(ZSTD_CCtx)`.

File: lib/compress/zstd_compress.c [hunk 3]
```diff
@@ -3191,32 +3210,29 @@ static size_t ZSTD_initCStream_stage2(ZSTD_CStream* zcs,
                                 unsigned long long pledgedSrcSize)
 {
     assert(!ZSTD_isError(ZSTD_checkCParams(params.cParams)));
+    zcs->blockSize = MIN(ZSTD_BLOCKSIZE_ABSOLUTEMAX, (size_t)1 << params.cParams.windowLog);
 
     /* allocate buffers */
-    {   size_t const neededInBuffSize = (size_t)1 << params.cParams.windowLog;
+    {   size_t const neededInBuffSize = ((size_t)1 << params.cParams.windowLog) + zcs->blockSize;
```
Comment (terrelln) [inline lib/compress/zstd_compress.c]: Why do we need an extra `zcs->blockSize` now?

File: lib/decompress/zstd_decompress.c [hunk 4]
```diff
@@ -105,7 +105,7 @@ struct ZSTD_DCtx_s
     const void* vBase;            /* virtual start of previous segment if it was just before current one */
     const void* dictEnd;          /* end of previous segment */
     size_t expected;
-    ZSTD_frameParams fParams;
+    ZSTD_frameHeader fParams;
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]: I like `ZSTD_frameHeader` as a name.

File: lib/decompress/zstd_decompress.c [hunk 5]
```diff
@@ -219,22 +219,23 @@ static size_t ZSTD_frameHeaderSize(const void* src, size_t srcSize)
 }
 
 
-/** ZSTD_getFrameParams() :
+/** ZSTD_getFrameHeader() :
 *   decode Frame Header, or require larger `srcSize`.
-*   @return : 0, `fparamsPtr` is correctly filled,
+*   @return : 0, `fhiPtr` is correctly filled,
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]: Where does the `i` come from?

File: lib/compress/zstd_compress.c (hunk 1 above)
Comment (Cyan4973) [inline lib/compress/zstd_compress.c]: Yes, they are now equivalent

File: lib/compress/zstd_compress.c (hunk 2 above)
Comment (Cyan4973) [inline lib/compress/zstd_compress.c]: Indeed, a copy/paste leftover ...

File: lib/decompress/zstd_decompress.c (hunk 5 above)
Comment (Cyan4973) [inline lib/decompress/zstd_decompress.c]: It used to be `ZSTD_frameHeaderInformation`, hence `fhi`.
It was later simplified to `ZSTD_frameHeader`.

I changed the name of this variable, it's now `zfhPtr`.

File: lib/compress/zstd_compress.c (hunk 3 above)
Comment (Cyan4973) [inline lib/compress/zstd_compress.c]: While not strictly necessary (it works without it),
it makes this policy a bit more favorable to unregular block sizes, improving compression ratio.

**Pull request merged**